torch_xla/experimental/checkpoint.py

```diff
@@ -0,0 +1,205 @@
+)
+from torch.distributed.checkpoint.planner import (
+    SavePlanner,
+    LoadPlanner,
+    SavePlan,
+    LoadPlan,
+    ReadItem,
+    WriteItem,
+    WriteItemType,
+    LoadItemType,
+)
+from torch.distributed.checkpoint.planner_helpers import (
+    create_read_items_for_chunk_list,)
+from torch.distributed.checkpoint.metadata import (
+    BytesStorageMetadata,
+    ChunkStorageMetadata,
+    TensorStorageMetadata,
+    MetadataIndex,
+    Metadata,
+    STATE_DICT_TYPE,
+    STORAGE_TYPES,
+)
+from torch.distributed.checkpoint._nested_dict import (
+    FLATTEN_MAPPING,
+    flatten_state_dict,
+)
+from torch.distributed.checkpoint.utils import find_state_dict_object
+from torch.utils._pytree import tree_map
+from torch_xla.experimental.xla_sharding import (XLAShardedTensor, XLAShard,
+                                                 ShardingType)
+from typing import List, Tuple, Dict
+
+__all__ = [
+    "XLASavePlanner",
+    "XLALoadPlanner",
+]
+
+
+class XLASavePlanner(SavePlanner):
+  """
+  XLASavePlanner provides an implementation of the SavePlanner interface
+  which handles state_dicts containing XLAShardedTensor.
+  """
+
+  def set_up_planner(self, state_dict: STATE_DICT_TYPE,
+                     is_coordinator: bool) -> None:
+    raise NotImplemented
+
+  def create_local_plan(self) -> SavePlan:
+    raise NotImplemented
+
+  def create_global_plan(
+      self, all_plans: List[SavePlan]) -> Tuple[List[SavePlan], Metadata]:
+    raise NotImplemented
+
+  def finish_plan(self, new_plan: SavePlan) -> SavePlan:
+    raise NotImplemented
+
+
+class XLALoadPlanner(LoadPlanner):
+  """
+  XLALoadPlanner provides an implementation of the LoadPlanner interface
+  which handles state_dicts containing XLAShardedTensor.
+
+  The input state_dict should already be sharded and on the XLA device, and
+  tensors and shards will be loaded in-place.
+  """
+
+  # Upon the first `resolve_tensor` call for a ReadItem associated with a
+  # sharded tensor, all local shards are moved to CPU via
+  # `XLAShardedTensor::local_shards` and are tracked in `_local_shards` until
+  # the last shard is committed in `commit_tensor`.
+  _local_shards: Dict[str, List[XLAShard]] = {}
+
+  # When `commit_tensor` is called, the updated XLAShard is tracked in
+  # `_processed_shards`, and when all shards have been committed, the
+  # XLAShardedTensor data is restored from the list here.
+  _processed_shards: Dict[str, List[XLAShard]] = {}
+
+  def set_up_planner(
+      self,
+      state_dict: STATE_DICT_TYPE,
+      metadata: Metadata,
+      is_coordinator: bool,
+  ) -> None:
+    self.metadata = metadata
+    self.is_coordinator = is_coordinator
+    self.original_state_dict = state_dict
+
+    # Flatten the state_dict to allow separating sharded XLA tensors from
+    # types that can be handled by the default planner, and ensure all sharded
+    # tensors are wrapped in XLAShardedTensor
+    state_dict, self.mappings = flatten_state_dict(state_dict)
```

Comment:
Hmm I actually don't know why it's optional in the DefaultLoadPlanner, I'll follow up on that. The flattened state_dict is easier for us to work with since we need to split the input into sharded/unsharded parts, which is difficult if the state_dict is nested.

One downside of flattening is that we aren't operating directly on the input state_dict, so non-tensor items need to be mapped back to the original state_dict.

Reply:
But that down side is handled by the default planner helpers, right?

Reply:
That's correct.

Reply:
It's optional on DefaultLoadPlaner mostly to help with testing. Its default value is True, so most users will always have it enabled.